api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,102 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+func (d *DeployProvider) Create(req models.CreateDeployRequest) (*models.Deploy, error) {
+	// TODO: see Line 59. Which entity ID generation pattern do we prefer? ARN to ID or the
+	// example below?
+	deployID := generateEntityID(req.DeployName)
+
+	if err := req.Validate(); err != nil {
+		return nil, err
+	}
+
+	deployFile, err := d.createRenderedDeploy(req.DeployFile)
+	if err != nil {
+		return nil, err
+	}
+
+	familyName := L0DeployID(d.Config.Instance(), req.DeployName)
+	if deployFile.Family != nil && deployFile.Family != aws.String(familyName) {
+		return nil, fmt.Errorf("Custom family names are currently unsupported in Layer0")
+	}
+
+	input := &ecs.RegisterTaskDefinitionInput{}
+	input.SetFamily(familyName)
+	input.SetTaskRoleArn(aws.StringValue(deployFile.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(deployFile.NetworkMode))
+	input.SetContainerDefinitions(deployFile.ContainerDefinitions)
+	input.SetVolumes(deployFile.Volumes)
+	input.SetPlacementConstraints(deployFile.PlacementConstraints)
+
+	taskDefinition := d.createTaskDefinition(input)
+
+	bytes, err := json.Marshal(taskDefinition)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract deploy file: %s", err.Error())
+	}
+
+	deploy := &models.Deploy{
+		DeployID:   L0DeployID(d.Config.Instance(), deployID),
+		Version:    GetRevision(d.Config.Instance(), deployID),
+		DeployFile: bytes,
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "name", Value: req.DeployName}); err != nil {
+		return deploy, err
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "version", Value: deploy.Version}); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createTaskDefinition(input *ecs.RegisterTaskDefinitionInput) *ecs.TaskDefinition {
```

Comment:
matching the return value from `renderTaskDefinition(body []byte) (*ecs.TaskDefinition, error)`, 
the argument for `createTaskDefinition` should probably be a `*ecs.Task` as well. 
That way, the code is pretty simple in the "Master" function, e.g.:
```
func CreateDeploy(...) {
    fqDeployID := ...
    ...
    taskDefinition, err := renderTaskDefinition(req.DeployFile)
    if err != nil {
        return nil, err
    }

    createdTask, err := d.createTaskDefinition(fqDeployID, taskDefinition)
    if err != nil {
        return nil, err
    }

   ...
}
```